src/glslplugin/annotation/impl/ArrayBoundAnnotation.java

```diff
@@ -0,0 +1,52 @@
+package glslplugin.annotation.impl;
+
+import com.intellij.lang.annotation.AnnotationHolder;
+import glslplugin.annotation.Annotator;
+import glslplugin.lang.elements.expressions.GLSLExpression;
+import glslplugin.lang.elements.expressions.GLSLSubscriptExpression;
+import glslplugin.lang.elements.types.GLSLArrayType;
+import glslplugin.lang.elements.types.GLSLType;
+import glslplugin.lang.elements.types.GLSLTypes;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Created by abigail on 08/07/15.
+ */
+public class ArrayBoundAnnotation extends Annotator<GLSLSubscriptExpression> {
+    @Override
+    public void annotate(GLSLSubscriptExpression expr, AnnotationHolder holder) {
+        GLSLExpression expression = expr.getArrayExpression();
+        GLSLExpression subscript = expr.getSubscript();
+        if (expression == null || subscript == null) return;
+        if (!expression.getType().isValidType() || !subscript.getType().isValidType()) return;
+
+        // It's valid for the expression not to be an array, as vectors and matrices support subscript syntax too.
+        // Either way, we'll handle ensuring the types are correct as a GLSLOperator
+        if (!(expression.getType() instanceof GLSLArrayType)) return;
```

Comment:
Vectors and matrices would benefit from constant-index-out-of-bounds checking as well. (`mat3(1.0)[-8]`) But that can wait.